clients/src/main/java/org/apache/kafka/common/telemetry/internals/KafkaMetricsCollector.java

```diff
@@ -0,0 +1,341 @@
+        if (!metricsEmitter.shouldEmitMetric(metricKey)) {
+            return;
+        }
+
+        metricsEmitter.emitMetric(
+            SinglePointMetric.sum(metricKey, value, true, clock.instant(), ledger.instantAdded(metricKey))
+        );
+    }
+
+    private void collectGauge(MetricKey metricKey, Number value, MetricsEmitter metricsEmitter) {
+        if (!metricsEmitter.shouldEmitMetric(metricKey)) {
+            return;
+        }
+
+        metricsEmitter.emitMetric(
+            SinglePointMetric.gauge(metricKey, value, clock.instant())
+        );
+    }
+
+    private static boolean isMeasurable(KafkaMetric metric) {
+        // KafkaMetric does not expose the internal MetricValueProvider and throws an IllegalStateException exception
+        // if .measurable() is called for a Gauge.
+        // There are 2 ways to find the type of internal MetricValueProvider for a KafkaMetric - use reflection or
+        // get the information based on whether or not a IllegalStateException exception is thrown.
+        // We use reflection so that we can avoid the cost of generating the stack trace when it's
+        // not a measurable.
+        try {
+            Object provider = METRIC_VALUE_PROVIDER_FIELD.get(metric);
+            return provider instanceof Measurable;
+        } catch (Exception e) {
+            throw new KafkaException(e);
+        }
+    }
+
+    /**
+     * Keeps track of the state of metrics, e.g. when they were added, what their getAndSet value is,
+     * and clearing them out when they're removed.
+     */
+    private class StateLedger {
+
+        private final Map<MetricKey, KafkaMetric> metricMap = new ConcurrentHashMap<>();
+        private final LastValueTracker<Double> doubleDeltas = new LastValueTracker<>();
+        private final Map<MetricKey, Instant> metricAdded = new ConcurrentHashMap<>();
+
+        private Instant instantAdded(MetricKey metricKey) {
+            // lookup when the metric was added to use it as the interval start. That should always
+            // exist, but if it doesn't (e.g. if there's a race) then we use now.
+            return metricAdded.computeIfAbsent(metricKey, x -> clock.instant());
+        }
+
+        public void init(List<KafkaMetric> metrics) {
+            log.info("initializing Kafka metrics collector");
+            for (KafkaMetric m : metrics) {
+                metricMap.put(metricNamingStrategy.metricKey(m.metricName()), m);
+            }
+        }
+
+        public void metricChange(KafkaMetric metric) {
+            MetricKey metricKey = metricNamingStrategy.metricKey(metric.metricName());
+            metricMap.put(metricKey, metric);
+            if (doubleDeltas.contains(metricKey)) {
+                log.warn("Registering a new metric {} which already has a last value tracked. " +
+                    "Removing metric from delta register.", metric.metricName(), new Exception());
```

Comment:
Why do we need to log a warning for this case? Seems it's expected (ie, nothing wrong) if we have a "last value" tracked for an existing metric?

`metricChange` is basically adding a new one, or updating an existing one. So why a WARN here?

Reply:
a metric should be removed before being added again, otherwise it's not clear if the underlying sensors have been reset or not, which makes it impossible to know if a delta is correct or not.

Reply:
Resolving the comment.